### ISSUE TITLE
No integer type has byte size <number>. Use a numeric with precision 0 instead.

### ISSUE THREAD

**Issue by mhluska** [zdennis/activerecord-import#559]
This library seems to fail with Rails 4.1 and `decimal` data types in Postgres. I'm getting this stack trace:

```sh
     Failure/Error:
       CoinPaymentTransaction.import(
         transactions,
         on_duplicate_key_update: {
           conflict_target: [:transaction_hash],
           columns: [:coin_conversion]
         }
       )

     ActiveRecord::ActiveRecordError:
       No integer type has byte size 24. Use a numeric with precision 0 instead.
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-4.1.16/lib/active_record/connection_adapters/postgresql/schema_statements.rb:478:in `type_to_sql'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-4.1.16/lib/active_record/connection_adapters/postgresql/quoting.rb:21:in `quote'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:932:in `block (2 levels) in values_sql_for_columns_and_attributes'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:911:in `each'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:911:in `each_with_index'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:911:in `each'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:911:in `map'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:911:in `block in values_sql_for_columns_and_attributes'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:910:in `map'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:910:in `values_sql_for_columns_and_attributes'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:763:in `import_without_validations_or_callbacks'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:721:in `import_with_validations'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:663:in `import_helper'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-import-0.27.0/lib/activerecord-import/import.rb:498:in `bulk_import'
     # ./lib/cryptocoin_payable/commands/payment_processor.rb:60:in `update_via_bulk_insert'
     # ./lib/cryptocoin_payable/commands/payment_processor.rb:43:in `block in update_transactions_for'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-4.1.16/lib/active_record/connection_adapters/abstract/database_statements.rb:201:in `block in transaction'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-4.1.16/lib/active_record/connection_adapters/abstract/database_statements.rb:209:in `within_new_transaction'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-4.1.16/lib/active_record/connection_adapters/abstract/database_statements.rb:201:in `transaction'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-4.1.16/lib/active_record/transactions.rb:208:in `transaction'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/activerecord-4.1.16/lib/active_record/transactions.rb:259:in `transaction'
     # ./lib/cryptocoin_payable/commands/payment_processor.rb:41:in `update_transactions_for'
     # ./spec/acceptance/commands/payment_processor_spec.rb:59:in `block (4 levels) in <top (required)>'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/benchmark-perf-0.4.0/lib/benchmark/perf.rb:98:in `clock_time'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/benchmark-perf-0.4.0/lib/benchmark/perf/execution_time.rb:66:in `block (2 levels) in run_warmup'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/benchmark-perf-0.4.0/lib/benchmark/perf/execution_time.rb:36:in `block in run_in_subprocess'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/benchmark-perf-0.4.0/lib/benchmark/perf/execution_time.rb:30:in `fork'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/benchmark-perf-0.4.0/lib/benchmark/perf/execution_time.rb:30:in `run_in_subprocess'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/benchmark-perf-0.4.0/lib/benchmark/perf/execution_time.rb:65:in `block in run_warmup'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/benchmark-perf-0.4.0/lib/benchmark/perf/execution_time.rb:64:in `times'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/benchmark-perf-0.4.0/lib/benchmark/perf/execution_time.rb:64:in `run_warmup'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/benchmark-perf-0.4.0/lib/benchmark/perf/execution_time.rb:87:in `run'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/rspec-benchmark-0.4.0/lib/rspec/benchmark/timing_matcher.rb:42:in `matches?'
     # ./spec/acceptance/commands/payment_processor_spec.rb:59:in `block (3 levels) in <top (required)>'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:123:in `block in run'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `loop'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `run'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/rspec-retry-0.6.1/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
     # /Users/mhluska/.rvm/gems/ruby-2.3.1/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:37:in `block (2 levels) in setup'
```

Versions:
activerecord 4.1.16
activerecord-import 0.27
ruby 2.3.1p112

Part of my schema:

```rb
ActiveRecord::Schema.define(version: 20_181_015_141_952) do
  create_table 'coin_payment_transactions', force: :cascade do |t|
    t.decimal 'estimated_value', precision: 24
    t.string 'transaction_hash'
    t.string 'block_hash'
    t.datetime 'block_time'
    t.datetime 'estimated_time'
    t.integer 'coin_payment_id'
    t.decimal 'coin_conversion', precision: 24
    t.integer 'confirmations', default: 0
    t.index ['coin_payment_id'], name: 'index_coin_payment_transactions_on_coin_payment_id'
    t.index ['transaction_hash'], name: 'index_coin_payment_transactions_on_transaction_hash', unique: true
  end
  ...
```

I'm trying to get Rails 4.1 support in Travis CI for https://github.com/Sailias/cryptocoin_payable/pull/20

**Comment by jkowens** [comment]
I think this might be an ActiveRecord quirk. Can you set a scale value along with precision and report back?

See: https://github.com/rails/rails/issues/17537

**Comment by mhluska** [comment]
I _think_ I tried setting the scale to zero a while back and it didn't help. Let me try again though.

**Comment by jkowens** [comment]
If I'm reading the above rails issue correctly, scale has to be >=1 for rails to interpret the field as a decimal. Otherwise it treats it like an integer.

**Comment by mhluska** [comment]
Hmm interesting. Let me look more into this and report back tomorrow. Or I don't mind closing this since it seems unrelated to activerecord-import.

**Comment by oniofchaos** [comment]
Does it fail with newer versions of Rails? Rails 4.1 is officially out of support by Rails.

**Comment by oniofchaos** [comment]
https://guides.rubyonrails.org/maintenance_policy.html#severe-security-issues

Fun upgrades ahead for you :D

**Comment by jkowens** [comment]
I'm closing this for now since it seems to be a Rails issue.
